x-pack/plugins/apm/public/new-platform/core_dependencies.ts

```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { CoreSetup } from 'src/core/public';
+
+export let basePath = {
+  addToPath: path => `/kbn-basepath${path}`
+} as CoreSetup['basePath'];
+
+// update the mutable singletons that enables access to core dependencies throughout the application
+// This could be passed down via React Context in the future
+// Consuming context from child components is easy with hooks, but hooks are only supported in functional components (not class components)
+// Also, non-react functions will not be able to access context - therefore singletons might not be the worst idea
+export function _setCoreDependencies(core: CoreSetup) {
+  basePath = core.basePath;
+}
```

Comment:
In the long run, I wonder if it'd be "more React-like" for us to work on converting imperative helper methods into declarative components that would then be able to use hooks to get the values from context.

Out of curiosity, how many imperative helpers do we currently have, and how many of them need access to these core deps? I imagine it's mostly the url helper stuff needing basePath, etc.

Reply:
The more I think about this, the more I think the `core` services should work the same as `location` -- seems weird to have `location` be accessible via context + hook (which also means you can't directly access it from imperative code) and then have these values cached in a singleton. Curious how you all are feeling about it, though?

Reply:
I think it makes sense to pass stateful objects (like `location`) down via context. But I find it a little odd to do the same with methods.

> core services should work the same as location

So should we use the singleton approach for everything, or React Context? If we use the latter, how should [callApi](https://github.com/elastic/kibana/blob/e44ad7b3586680efcd35a1e39cc5955f52388482/x-pack/plugins/apm/public/services/rest/callApi.ts#L49) access `kfetch`?

Reply:
Hm, yeah. My first thought is that if `kfetch` is no longer globally stateful (i.e. the platform isn't setting things on it before providing it to the plugin), then it should just be a separate package we import via npm/yarn. Anything else that's just a stateless method should work the same way, imo, then the rest would be handled via React context. I think we'll need to talk to the platform folks about this...

@weltenwort what is your team thinking wrt to how to handle things like this in the new platform?

Reply:
I'd definitely importing it statically if that's an option!

Reply:
While I can't speak for the whole infra monitoring team, I've started to write hooks for such external services (e.g. `useKibanaUiSettings('mySetting')`) in https://github.com/elastic/kibana/pull/34615/files#diff-5c1aef9186a64f50ff1ef32ee57a92ccR29) to abstract these platform details away.

These hooks can then either get the reference from the context if it needs to be injected or statically import it if possible.

Reply:
That way the plugin dependencies could be injected with one `platformDependencies` provider, but fine-grained hooks hide the fact that we're extracting the individual component from one larger object. Not sure if I'm making sense, let me know if you want to discuss it in person after I had my morning tea. :wink: